lib/bot_framework/session.rb

```diff
@@ -0,0 +1,422 @@
+      send_localized(args,message)
+    end
+
+    def send_localized(localization_namspace,message,args=[])
+      #TODO Incomplete
+      @msg_sent = true
+      m = {text: message}
+      prepare_message(m)
+      @batch << m
+      self
+    end
+
+    # Sends a typing indicator to the user
+    def send_typing
+      @msg_sent = true
+      m = { type: 'typing' }
+      m = prepare_message(m)
+      @batch.push(m)
+      logger.info 'Send Typing'
+      send_batch 
+    end
+
+    def send_batch
+      logger.info "Sending batch with elements #{@batch.count}"
+      return if sending_batch?
+      # TODO: timer
+      @batch_timer = nil
+      batch = @batch
+      @batch_started = false
+      @sending_batch = true
+      cur = cur_dialog
+      cur.state = @dialog_data
+      options[:on_save] = proc do |err|
+        if err
+          @sending_batch = false
+          case (err.code || '')
+          when 'EBADMSG'
+          when 'EMSGSIZE'
+            # Something went wrong , let's reset everything
+            @user_data = {}
+            @batch = []
+            end_conversation(@options[:dialog_error_message] || 'Something went wrong and we need to startover')
+          end
+          yield(err) if block_given?
+        else
+          if batch.length
+            options[:on_send].call(batch) do |err|
+              @sending_batch = false
+              start_batch if @batch_started # What is this :o
+              yield(err) if block_given?
+            end
+          else
+            @sending_batch = false
+            start_batch if @batch_started
+            yield(err) if block_given?
+          end
+        end
+      end
+    end
+
+    
+    # Begin a new dialog
+    def begin_dialog(id, args = nil)
+      logger.info "Beginning new dialog #{id}"
+      id = resolve_dialog_id(id)
+      dialog = find_dialog(id)
+      raise "Dialog #{id} Not found" unless dialog
+      push_dialog(id: id, state: {})
+      start_batch
+      dialog.begin(self, args)
+      self
+    end
+
+    def replace_dialog(id, args = nil)
+      logger.info "Session replace dialog #{id}"
+      id = resolve_dialog_id(id)
+      dialog = find_dialog(id)
+      raise "Dialog #{id} Not found" unless dialog
+
+      # Update the stack and start dialog
+      pop_dialog
+      push_dialog(id: id, state: {})
+      start_batch
+      dialog.begin(self, args)
+      self
+    end
+
+    # End conversation with the user
+    def end_conversation(message = nil, _args = {})
+      if message
+        # TODO: sent message
+      end
+      # Clear private conversation data
+      @private_conversation_data = {}
+
+      # Clear stack and save
+      logger.info 'End conversation'
+      ss = @session_state
+      ss.call_stack = []
+      send_batch
+      self
+    end
+
+    def end_dialog(message = nil, args = {})
+      if message
+        # TODO: end_dialog_with_result
+      end
+      cur = cur_dialog
+      if cur
```

Comment:
Use a guard clause instead of wrapping the code inside a conditional expression.